appscale/tools/remote_helper.py

```diff
@@ -940,41 +942,46 @@ def terminate_virtualized_cluster(cls, keyname, is_verbose):
 
     acc = AppControllerClient(shadow_host, secret)
     try:
-      all_ips = acc.get_all_public_ips()
+      terminate_thread = threading.Thread(target=acc.run_terminate,
+                                          args=(clean,))
+      terminate_thread.start()
+      terminated_successfully = True
+      log_dump = ""
+      while not acc.is_appscale_terminated():
+        # for terminate receive_server_message will return a JSON string that
+        #  is a list of dicts with keys: ip, status, output
+        output_list = yaml.safe_load(acc.receive_server_message())
+        if hasattr(output_list, "startswith") and \
+            output_list.startswith("Error"):
+          log_dump += output_list
+          continue
+        for node in output_list:
+          if node.get("status"):
+            AppScaleLogger.success("Node at {node_ip}: {status}".format(
+              node_ip=node.get("ip"), status="Terminated Successfully"))
+          else:
+            AppScaleLogger.warn("Node at {node_ip}: {status}".format(
+              node_ip=node.get("ip"), status="Did not terminate successfully"))
+            terminated_successfully &= False
+            log_dump += "Node at {node_ip}: {status}\nNode Output:"\
+                        "{output}".format(node_ip=node.get("ip"),
+                                          status="Terminate failed",
+                                          output=node.get("output"))
+          AppScaleLogger.verbose("Output of node at {node_ip}:\n"\
+                                 "{output}".format(node_ip=node.get("ip"),
+                                                   output=node.get("output")),
+                                 is_verbose)
+      terminate_thread.join()
+      if not terminated_successfully:
+        AppScaleLogger.warn("Some nodes failed terminating")
+        LocalState.generate_crash_log(AppControllerException, log_dump)
+      cls.stop_remote_appcontroller(shadow_host, keyname, is_verbose)
```

Comment:
Unfortunately, that would mean that there is no easy way for the user to retry the terminate, and leaving processes running on other nodes leads to pretty nasty side effects on subsequent ups.

> If all the nodes fail to come down there is something bigger wrong and we would not be able to run terminate the next time either.

Why wouldn't we be able to run a terminate again if the shadow's controller is still running?